streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java

```diff
@@ -0,0 +1,351 @@
+            //Only one among kafkaStreams1 and kafkaStreams2 will contain the specific active store requested. The other will throw exception
+        }
+        try {
+            store4 = kafkaStreams2
+                    .store(StoreQueryParams.fromNameAndType(TABLE_NAME, queryableStoreType).withPartition(keyDontBelongPartition));
+        } catch (final InvalidStateStoreException exception) {
+            //Only one among kafkaStreams1 and kafkaStreams2 will contain the specific active store requested. The other will throw exception
+        }
+
+        // Assert that key is not served when wrong specific partition is requested
+        // If kafkaStreams1 is active for keyPartition, kafkaStreams2 would be active for keyDontBelongPartition
+        // So, in that case, store3 would be null and the store4 would not return the value for key as wrong partition was requested
+        assertThat(kafkaStreams1IsActive ? store4.get(key) : store3.get(key), is(nullValue()));
+    }
+
+    @Test
+    public void shouldQueryAllStalePartitionStores() throws Exception {
+        final int batch1NumMessages = 100;
+        final int key = 1;
+        final Semaphore semaphore = new Semaphore(0);
+
+        final StreamsBuilder builder = new StreamsBuilder();
+        builder
+                .table(INPUT_TOPIC_NAME, Consumed.with(Serdes.Integer(), Serdes.Integer()),
+                        Materialized.<Integer, Integer, KeyValueStore<Bytes, byte[]>>as(TABLE_NAME)
+                                .withCachingDisabled())
+                .toStream()
+                .peek((k, v) -> semaphore.release());
+
+        final KafkaStreams kafkaStreams1 = createKafkaStreams(builder, streamsConfiguration());
+        final KafkaStreams kafkaStreams2 = createKafkaStreams(builder, streamsConfiguration());
+        final List<KafkaStreams> kafkaStreamsList = Arrays.asList(kafkaStreams1, kafkaStreams2);
+
+        startApplicationAndWaitUntilRunning(kafkaStreamsList, Duration.ofSeconds(60));
+
+        produceValueRange(key, 0, batch1NumMessages);
+
+        // Assert that all messages in the first batch were processed in a timely manner
+        assertThat(semaphore.tryAcquire(batch1NumMessages, 60, TimeUnit.SECONDS), is(equalTo(true)));
+
+        final QueryableStoreType<ReadOnlyKeyValueStore<Integer, Integer>> queryableStoreType = QueryableStoreTypes.keyValueStore();
+        final ReadOnlyKeyValueStore<Integer, Integer> store1 = kafkaStreams1
+                .store(StoreQueryParams.fromNameAndType(TABLE_NAME, queryableStoreType).enableStaleStores());
+
+        final ReadOnlyKeyValueStore<Integer, Integer> store2 = kafkaStreams2
+                .store(StoreQueryParams.fromNameAndType(TABLE_NAME, queryableStoreType).enableStaleStores());
+
+        // Assert that both active and standby are able to query for a key
+        assertThat(store1.get(key), is(notNullValue()));
+        assertThat(store2.get(key), is(notNullValue()));
```

Comment:
Seem this test is subject to a race condition, because we don't know if the standby is up-to-date already (the used semaphore only tells us that the active one is up-to-date) and contains the key. I think we need to retry this in a loop with timeout?

Reply:
Done.